### PR TITLE
fix: save rescanning progress

### PR DESCRIPTION
**Pull request by sander2**
To address https://github.com/interlay/interbtc-clients/issues/471 .

Note that I went with another approach - rather than storing data in a hacky way inside the bitcoin wallet, I store it in a rocksdb (since we'll be stateful anyway with https://github.com/interlay/interbtc-clients/pull/246). I wrote it in such a way that the database can be removed without any bad side-effects: it will just rescan everything again. This means that users don't necessarily have to copy over the database when migrating their vaults to other servers

### CODE REVIEW

File: vault/src/issue.rs [hunk 1]
```diff
@@ -135,6 +188,30 @@ pub async fn add_keys_from_past_issue_request(
             .await?;
     }
 
+    // save progress s.t. we don't rescan pruned range again if we crash now
+    scanning_status.store(vault_id, db)?;
+
+    let mut chunk_size = 1;
```
Comment (gregdhill) [inline vault/src/issue.rs:194]: Should this not be higher?

Comment (gregdhill) [reply vault/src/issue.rs:194]: Oh it changes depending on how long the rescanning took, that makes sense.

File: vault/src/issue.rs [hunk 2]
```diff
@@ -374,3 +451,185 @@ pub async fn listen_for_issue_cancels(
...
+    //         let (start, end) = self.queued_rescan_range?;
+    //         let chunk_end = end.min(start + max_blocks);
+    //
+    //         if chunk_end == end {
+    //             self.queued_rescan_range = None; // this will be the last chunk to scan
+    //         } else {
+    //             self.queued_rescan_range = Some((chunk_end + 1, end));
+    //         }
+    //         Some((start, chunk_end))
+    //     }
```
Comment (gregdhill) [inline vault/src/issue.rs:634]: Is this still needed?

File: vault/tests/vault_integration_tests.rs [hunk 3]
```diff
@@ -126,17 +131,25 @@ async fn test_replace_succeeds() {
...
         let btc_rpcs = vec![(new_vault_id.clone(), btc_rpc.clone())].into_iter().collect();
         let new_btc_rpc_master_wallet = btc_rpc.clone();
-        let _vault_id_manager =
-            VaultIdManager::from_map(new_vault_provider.clone(), new_btc_rpc_master_wallet, btc_rpcs);
+        let _vault_id_manager = VaultIdManager::from_map(
+            new_vault_provider.clone(),
+            new_btc_rpc_master_wallet,
+            btc_rpcs,
+            "test_replace_succeeds1",
+        );
```
Comment (gregdhill) [inline vault/tests/vault_integration_tests.rs:139]: Doesn't look like we need this, or is the construction required?

Comment (sander2) [reply vault/tests/vault_integration_tests.rs:139]: It's just to ensure that tests don't interfere with each other by reading eath other's databases. We could use something like tempdir but I think for now this is fine

File: vault/src/issue.rs [hunk 4]
```diff
@@ -135,6 +188,30 @@ pub async fn add_keys_from_past_issue_request(
...
+        let start_time = Instant::now();
+
+        bitcoin_core.rescan_blockchain(chunk_start, chunk_end).await?;
+
+        // with the code below the rescan time should remain between 5 and 20 seconds
+        // after the first couple of rounds.
+        if start_time.elapsed() < Duration::from_secs(10) {
+            chunk_size = chunk_size.saturating_mul(2);
+        } else {
+            chunk_size = (chunk_size / 2).max(1);
```
[the remark this answers is not in the document]
Comment (sander2) [reply vault/src/issue.rs:209]: It can't fail, but sure

**Review by gregdhill — APPROVED** [review]

**Pull request merged**